pkg/fetcher/fetcher.go

```diff
@@ -414,6 +428,67 @@ func (u *URLFetcher) IsNonretryableClientError() bool {
 		s != http.StatusLocked && s != http.StatusTooManyRequests
 }
 
+// buildRegistryErrMsg builds error message for unexpected http code (nonretryable client errors and all other errors)
+// and extracts message details from response body stream if there is one (#5951).
+func (u *URLFetcher) buildRegistryErrMsg(url *url.URL, respBody io.ReadCloser) string {
+	errMsg := fmt.Sprintf("Unexpected http code: %d (%s), URL: %s", u.StatusCode, http.StatusText(u.StatusCode), url)
+
+	errDetail, err := extractErrResponseMessage(respBody)
+	if err != nil {
+		return errMsg
+	}
+
+	errMsg += fmt.Sprintf(", Message: %s", errDetail)
+	return errMsg
+}
+
+// extractErrResponseMessage extracts `message` field from error response body stream.
+func extractErrResponseMessage(rdr io.ReadCloser) (string, error) {
+	defer rdr.Close()
+
+	out := bytes.NewBuffer(nil)
+	_, err := io.Copy(out, rdr)
+	if err != nil {
+		log.Debugf("Error when copying from error response body stream: %s", err)
+		return "", err
+	}
+
+	res := []byte(out.Bytes())
+	log.Debugf("Error message json string: %s", string(res))
+
+	var errResponse RegistryErrorRespBody
+	err = json.Unmarshal(res, &errResponse)
+	if err != nil {
+		log.Debugf("Error when unmarshaling error response body: %s", err)
+		return "", err
+	}
+
+	if len(errResponse.Errors) == 0 {
+		log.Debugf("Error response wrong format. Response body: %s", string(res))
+		return "", fmt.Errorf("error response json has unconventional format")
```

Comment:
After reviewing the unit test: since `fmt.Errorf("error response json has unconventional format")` is used in multiple places, you can define a variable in this file like so: `var errJsonFormat = fmt.Errorf("error response json has unconventional format")` and use it here and in the unit test.